### PR TITLE
Enable unloading of AssemblyLoadContext

### PR DESCRIPTION
**Pull request by janvorli**
This change is basically the work of @xoofx done some time ago, rebased
to the current master and with added fixes. 

My contributions to it were:
* Extensive testing - running all 11753 coreclr Pri1 tests loaded into
AssemblyLoadContext including their dependencies and ensuring 
that they work (except the ones that use features not yet supported 
for unloading, like thread local members, COM interop).
That uncovered issues listed below.
* Fix few memory leaks
* Enable unloading of assemblies with PInvokes
* Fix virtual stub manager initialization
* Fix issue with unwind info regions reporting for stub linker
* Fix misplaced assert related to statics in collectible assemblies
* Fix issue with SOS domain enumeration.
* Fixed tests that were allocating, but not freeing GCHandle that
was preventing the tests from unloading.

I recommend viewing the commits separately.

There is still a lot of work to do to enable e.g. unloading of assemblies
with classes having thread local members and other stuff. But it seems
worth merging these changes in.

### CODE REVIEW

**Comment by janvorli** [comment]
CC: @xoofx, @sergiy-k

File: tests/src/JIT/Directed/nullabletypes/Desktop/StructDefinitions.cs [hunk 1]
```diff
@@ -236,6 +236,11 @@ static Helper()
         GCHANDLE = GCHandle.Alloc(Console.Out);
     }
 
+    public static void Shutdown()
```
Comment (jkotas) [inline tests/src/JIT/Directed/nullabletypes/Desktop/StructDefinitions.cs]: It would be better for this `Helper` to create object with finalizer to clean this up or register unload event to clean this up, rather than changing hundreds of tests to call the Shutdown method manually.

**Comment by xoofx** [comment]
Thanks @janvorli for bringing this back! And very happy to see also that "_Enable unloading of assemblies with PInvokes_"  ❤️

**Comment by janvorli** [comment]
@jkotas I've updated the commit that fixes the test unloadability according to your feedback.

**Comment by jkotas** [comment]
I have written a small test so I can step through some of the code. I am having troubles to make it unload properly. E.g. if I run the following - the memory will just keep growing (I have killed it when the working set was 2GB, on Windows x64 checked):

```csharp
using System;
using System.Threading.Tasks;
using System.Reflection;
using System.Runtime.Loader;

class MyLoadContext : AssemblyLoadContext
{
    public MyLoadContext() : base(isCollectible: true)
    {
    }

    protected override Assembly Load(AssemblyName assemblyName)
    {
        return null;
    }

    static void Work()
    {
        for (;;)
        {
            var alc = new MyLoadContext();
            alc.Unload();
        }
    }

    static void Main()
    {
        for (int i = 0; i < 10; i++) Task.Run(new Action(Work));
        Work();
    }
}
```

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 2]
```diff
@@ -128,37 +226,89 @@ public Assembly LoadFromStream(Stream assembly, Stream assemblySymbols)
...
+            {
+                throw new InvalidOperationException(SR.GetResourceString("AssemblyLoadContext_Unload_AlreadyUnloaded"));
             }
+            state = InternalState.Unloading;
+        }
 
-            return loadedAssembly;
+        private void VerifyIsAlive()
+        {
+            if (state != InternalState.Alive)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: When the unload event handlers are executing, their JITing may trigger loads of more assemblies into the AssemblyLoadContext. But this block will prevent those assemblies from being loaded because the state is not Alive anymore. I think we need to allow loading of more stuff into the assembly load context even when the unload event handlers are executing.

We should have a test for this.

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 3]
```diff
@@ -128,37 +226,89 @@ public Assembly LoadFromStream(Stream assembly, Stream assemblySymbols)
...
+                            new IntPtr(ptrSymbols), iSymbolLength, JitHelpers.GetObjectHandleOnStack(ref loadedAssembly));
+                    }
+                }
+                return loadedAssembly;
+#if FEATURE_COLLECTIBLE_ALC
+            }
+#endif
+        }
+
+        public void Unload()
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: I forgot what contract for this method we agreed on: Is it required to call this method to unload the stuff? Or is the ALC going to auto-unload itself too when these are no references to it?

It may be worth a comment.

**Comment by masonwheeler** [comment]
@jkotas 
> Edit: It looks like that there is a scalability problem. It does collect when I sprinkle it with sleeps. Still, it does not look right that it is possible to create these faster than the system is able to get rid of them.

This isn't particularly surprising, considering that the amount of work that has to be done to verify a context is safe to unload is greater than the amount of work that has to be done to create one, and that (I can only assume, because if not then none of this makes any sense in the first place) it's happening asynchronously on a thread that occasionally has other things to do, while this is simply spinning forever creating contexts and dumping them as quickly as possible.

This doesn't feel like something that would happen in real-world usage.  We'd definitely never see it with file-based assemblies, and even a script engine that's generating assemblies dynamically wouldn't approximate this level of turnover, because there would be space in between where it's compiling and then executing its generated code.

So basically what you've shown is that it's possible for deliberately malicious code to launch a DOS attack against this system, but that's a bit redundant because the hypothetical malicious code could be even more effective by not calling `Unload` in the first place.  That being true, I don't think this scenario is something to worry about in and of itself.

**Comment by jkotas** [comment]
> This doesn't feel like something that would happen in real-world usage

ok

File: src/vm/assemblynative.cpp [hunk 4]
```diff
@@ -1299,7 +1304,46 @@ INT_PTR QCALLTYPE AssemblyNative::InitializeAssemblyLoadContext(INT_PTR ptrManag
     {
         // Initialize a custom Assembly Load Context
         CLRPrivBinderAssemblyLoadContext *pBindContext = NULL;
+
+        // Create a new AssemblyLoaderAllocator for an AssemblyLoadContext
+        AssemblyLoaderAllocator* loaderAllocator = new AssemblyLoaderAllocator();
```
Comment (jkotas) [inline src/vm/assemblynative.cpp]: We should need to do this for regular non-unloadable load context. It is unnecessary overhead for them.

File: CMakeLists.txt [hunk 5]
```diff
@@ -584,6 +584,7 @@ add_subdirectory(src/pal/prebuilt/inc)
 
 add_subdirectory(src/debug/debug-pal)
 
+add_definitions(-DFEATURE_COLLECTIBLE_ALC)
```
Comment (jkotas) [inline CMakeLists.txt]: I am not sure whether we need this define. It made sense when this started and the same code was still building for desktop. It is not the case anymore.

File: src/vm/assemblyspec.cpp [hunk 6]
```diff
@@ -1560,7 +1559,12 @@ BOOL AssemblySpecBindingCache::StoreAssembly(AssemblySpec *pSpec, DomainAssembly
     UPTR key = (UPTR)pSpec->Hash();
 
     // On CoreCLR, we will use the BinderID as the key 
-    ICLRPrivBinder* pBinderContextForLookup = pAssembly->GetFile()->GetBindingContext();
+    ICLRPrivBinder* pBinderContextForLookup = pSpec->GetBindingContext();
```
Comment (jkotas) [inline src/vm/assemblyspec.cpp]: Why can't we just use `pAssembly->GetFile()->GetBindingContext()` as before?

**Comment by jkotas** [comment]
> There is still a lot of work to do to enable e.g. unloading of assemblies with classes having thread local members and other stuff.

And adding tests...

**Comment by jkotas** [comment]
LGTM modulo comments.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by xoofx** [comment]
> Edit: It looks like that there is a scalability problem. It does collect when I sprinkle it with sleeps. Still, it does not look right that it is possible to create these faster than the system is able to get rid of them.

Yeah, but I remember the problem was also on the original PR but afair, the ALC are delayed collected via `SystemDomain::ProcessDelayedUnloadDomains` which is indirectly trigerred via the finalizer of thread worker (via `FinalizerThread::FinalizerThreadWorker`), so unlikely to happen in the pathological case you described. So I agree with @masonwheeler.

**Comment by jakobbotsch** [comment]
How long would those sleeps need to be? [Fuzzlyn](https://github.com/jakobbotsch/Fuzzlyn) loads a lot of assemblies, and because of no app domains in .NET core, it launches a new process for every 100 assemblies to load them. With unloadable `AssemblyLoadContext` I assume the new processes could be avoided? I just measured and it amounts to around 45 assemblies loaded per second.

**Comment by xoofx** [comment]
> How long would those sleeps need to be?

There is something weird I can't remember exactly... but I'm vaguely remembering that there was a code that was actually called on some IDLE event happening on the app domain... but now can't find this code anymore... I also recall to have a brute force simple app, creating a bunch of ALC and unload them directly, it was not increasing in memory (even though there was two cases, one that hasn't loaded anything and one that loaded already something... maybe I tested only one that tested with at least one assembly loaded through it)...  That's difficult to recall all the details after more than 1.5 years, I should have put more notes around :sweat:

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
Looks like one of the System.Reflection.Emit.Tests failed with assert in the runtime that I was not hitting locally. I'm looking into it.

**Comment by masonwheeler** [comment]
@janvorli A wild WOMM appears!

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
I've made a couple of changes, some reflecting the feedback and some based on issues with Unloading event I've found:
* Remove FEATURE_CORECLR usage
* Remove SuppressUnmanagedCodeSecurity attribut
* Fix indentation (there were tabs at few place
* Add missing error messages to the resources
* Remove FEATURE_COLLECTIBLE_ALC define
* Don't allocate AssemblyLoaderAllocator for non-unloadable context
* Fix unloading of some tests in using Unloading event to clean up GCHandle they were allocating
* Moved the Unloading event raising to the Unload method and the AssemblyLoadContext finalizer to enable the code running inside the context to perform cleanup as soon as possible. There was a problem that when an event handler was registered, it held everything alive, since the underlying delegate keeps LoaderAllocator reference. So the finalizer of LoaderAllocatorScout was never called as long as the event had something registered.
* Thanks to the previous change, I could remove the transformation of the m_ptrManagedAssemblyLoadContext handle from weak to strong in the `CLRPrivBinderAssemblyLoadContext::PrepareForLoadContextRelease`.
* Refactored unloading that was done little differently in the AssemblyLoadContext finalizer, the Unload method and the OnAppContextUnloading into single common function.
* Stopped registering non-collectible AssemblyLoadContext instances in the ContextsToUnload dictionary.

**Comment by masonwheeler** [comment]
@janvorli This is awesome! We all really appreciate the work you're putting into this; this is a feature that's been sorely needed since forever.

**Comment by janvorli** [comment]
> When the unload event handlers are executing, their JITing may trigger loads of more assemblies into the AssemblyLoadContext. But this block will prevent those assemblies from being loaded because the state is not Alive anymore. I think we need to allow loading of more stuff into the assembly load context even when the unload event handlers are executing.

@jkotas, with the latest changes, it is not a problem anymore. The AssemblyLoadContext is still alive when the event is called, we switch the state to unloading after that.

**Comment by janvorli** [comment]
> I forgot what contract for this method we agreed on: Is it required to call this method to unload the stuff? Or is the ALC going to auto-unload itself too when these are no references to it?

Both ways work now.

**Comment by janvorli** [comment]
17 Windows NT arm tests are failing due to an AV at an unknown location. 
I need to look into that.

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test please

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by masonwheeler** [comment]
> shuffle thunk cache



What in the world is a shuffle thunk cache?

**Comment by janvorli** [comment]
> What in the world is a shuffle thunk cache?

When you use a delegate to call a method, the JIT doesn't know at the time it generates the code what the delegate points to. It can e.g. be a member method or a static method. So the JIT generates arguments to registers and stack based on the signature of the delegate and the call then doesn't call the target method directly, but a shuffle thunk instead. This thunk is generated based on the caller side signature and the real target method signature and shuffles the arguments in registers and on stack to correspond to the target calling convention. So if it needs to add "this" pointer into the first argument register, it needs to move the first argument register to the second, the second to the third and the last to the stack (obviously in the right order so that nothing gets overwritten). And e.g. Unix amd64 calling convention makes it even more interesting when there are arguments that are structs that can be passed in multiple registers.

**Comment by janvorli** [comment]
I've forgotten to add that the shuffle thunk then tail calls the real target method.

**Comment by masonwheeler** [comment]
Hmm... I suppose that makes sense if you accept that particular delegate model, but it still seems kind of strange.

Why not do this instead?

 * the delegate calls exactly what is given to it
 * when giving an instance method to a delegate nothing special has to be done
 * when giving a static method to a delegate, the JIT special-cases this to generate a shuffle thunk for the static method as a "magic" method on the same class, and assign that to the delegate instead

I would intuitively expect that giving an instance method to a delegate is by far the most common case, and that this model would greatly simplify the low-level implementation of use of delegates and lead to at least a modest performance improvement.

**Comment by jkotas** [comment]
> Why not do this instead?

This is how it works. The only difference is that the shuffle thunk is hand-emitted in assembly and not generated by the JIT for historic reasons. All stubs (including all interop stubs) were hand-emitted like this in .NET Framework 1.0. Starting with .NET Framework 2.0, we have been converting the stubs to be generated by the JIT (the runtime generates IL for the stub, and then the JIT compiles the IL as regular method). The shuffle thunk is one of the few remaining ones not converted yet. Also, we have the IL path on some platforms but not others - `FEATURE_STUBS_AS_IL` is related to it.

**Comment by masonwheeler** [comment]
That makes sense, thanks!

**Comment by janvorli** [comment]
@jkotas it looks I have nailed all the issues discovered by the coreclr and corefx tests and reflected all of your feedback from the first round of review. Could you please take another look now?

The Ubuntu x64 corefx test are systematically failing due to an unrelated issue: 
`The "IsRestoreRequired" task could not be loaded from the assembly /mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/_/fx/Tools/Microsoft.DotNet.Build.Tasks.dll`

**Comment by jkotas** [comment]
Closing & reopen to pick up a new CI definition

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by janvorli** [comment]
@jkotas the `Ubuntu x64 Checked Build and Test (Jit - CoreFx)` still fails with the same issue in Microsoft.DotNet.Build.Tasks.dll.

**Comment by jkotas** [comment]
Looks like #18498. @BruceForstall Is there a way to workaround it?

**Comment by BruceForstall** [comment]
I don't think it's that issue. I see:

https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_ubuntu_corefx_baseline_prtest/364/consoleFull
```
16:52:45 Error compiling /mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/Tools/Microsoft.DotNet.Build.Tasks.dll: Could not load file or assembly 'Microsoft.Build.Framework, Version=14.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
16:52:45  (Exception from HRESULT: 0x80070002)
16:52:45 Error: file "/mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/Tools/Microsoft.DotNet.Build.Tasks.dll" or one of its dependencies was not found
16:52:45 Unable to successfully compile /mnt/j/workspace/dotnet_coreclr/master/jitstress/x64_checked_ubuntu_corefx_baseline_prtest/Tools/Microsoft.DotNet.Build.Tasks.dll
```

With #18498 the build would succeed, but the tst and flow jobs would fail.

File: src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs [hunk 7]
```diff
@@ -32,37 +50,90 @@ public abstract class AssemblyLoadContext
...
+            InitiateUnload();
+        }
 
-            // Initialize event handlers to be null by default
-            Resolving = null;
+        private void InitiateUnload()
+        {
+            var unloading = Unloading;
             Unloading = null;
+            unloading?.Invoke(this);
```
Comment (davidwrighton) [inline src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs]: This may be launching code inside of the finalizer thread. Do we think that's ok?

File: src/vm/assemblyspec.cpp (hunk 6 above)
Comment (davidwrighton) [reply src/vm/assemblyspec.cpp]: It doesn't seem unreasonable that there may be a bug here, but I don't think its related to unloadability. This change is plenty complicated without adding possibly unrelated changes to it. @janvorli could you look into if this is necessary for the scenarios which you are testing?

File: src/vm/assemblyspec.cpp [hunk 8]
```diff
@@ -1660,9 +1672,29 @@ BOOL AssemblySpecBindingCache::StoreFile(AssemblySpec *pSpec, PEAssembly *pFile)
...
+            SafeComHolder<ICollectibleAssemblyLoadContext> pAssemblyLoadContext = NULL;
+            if (SUCCEEDED(pBinderContextForLookup->QueryInterface<ICollectibleAssemblyLoadContext>(&pAssemblyLoadContext)))
+            {
+                pAssemblyLoadContext->GetLoaderAllocator(&pLoaderAllocator);
+                _ASSERTE(pLoaderAllocator != NULL);
+                pHeap = pLoaderAllocator->GetHighFrequencyHeap();
+            }
+        }
+#endif // !CROSSGEN_COMPILE
+
```
Comment (davidwrighton) [inline src/vm/assemblyspec.cpp]: These changes are effectively creating a binding from an assemblySpec which lives in one loaderallocator to the loaderallocator of the collectible assembly. I don't see the associated logic that would add the the loaderallocator graph so that we maintain lifetime correctly, in the presence of interrelated collectible assembly load contexts

File: src/vm/assemblyspec.cpp [hunk 9]
```diff
@@ -1767,6 +1799,40 @@ BOOL AssemblySpecBindingCache::StoreException(AssemblySpec *pSpec, Exception* pE
     }
 }
 
+BOOL AssemblySpecBindingCache::RemoveAssembly(DomainAssembly* pAssembly)
```
Comment (davidwrighton) [inline src/vm/assemblyspec.cpp]: This appears to be logic for removing a DomainAssembly from the AssemblySpecBindingCache when it is unloaded, but I don't see the case where AssemblyLoadContext 1 has a reference to an assembly in AssemblyLoadContext 2, and we unload AssemblyLoadContext 1.

Comment (janvorli) [reply src/vm/assemblyspec.cpp]: I believe this should be ok. The `RemoveAssembly` is called from `LoaderAllocator::GCLoaderAllocators_RemoveAssemblies` and we only call it for assemblies that are loaded in the `AssemblyLoadContext` that's being destroyed. They are taken from the `m_pFirstDomainAssemblyFromSameALCToDelete` list.

File: src/vm/assemblyspec.cpp (hunk 6 above)
Comment (janvorli) [reply src/vm/assemblyspec.cpp]: Reverting this change doesn't have any impact on my testing scenarios.

File: src/vm/assemblyspec.cpp (hunk 8 above)
Comment (janvorli) [reply src/vm/assemblyspec.cpp]: Similar to the AssemblySpecBindingCache::StoreAssembly, I've tried to modify the way we get the binder context to use the context from the pFile and not try to use the context from the pSpec. @davidwrighton, it seems this should remove the issue, right?

Comment (davidwrighton) [reply src/vm/assemblyspec.cpp]: ? That sounds like a semantic change to assembly loading. The context that we use is a critical thing, and has implications for future binds. Lets talk about that.

File: src/vm/assemblyspec.cpp (hunk 9 above)
Comment (davidwrighton) [reply src/vm/assemblyspec.cpp]: That's precisely my point. That isn't covering all the cases in which you should need to delete things.

**Comment by janvorli** [comment]
I just got back from my long vacation and I'll continue working on this.

**Comment by janvorli** [comment]
@davidwrighton, as we have already discussed offline, I've looked into your concerns related to the binder cache. When an AssemblyLoadContext.Load  override gets an Assembly by calling another context's LoadFromAssemblyName name, it gets cached only for this other context. And this AssemblyLoadContext is held alive by the RuntimeAssembly's m_syncRoot. 

I've also reverted back the way we get the binder for lookup in the AssemblySpecBindingCache::StoreAssembly and AssemblySpecBindingCache::StoreFile based on the feedback - my testing didn't show a need for that.

**Comment by janvorli** [comment]
I've apparently made some mistake when removing my instrumentation logging that I have added for my personal debugging purposes, that's why the tests are failing. I'm looking into it.

**Comment by janvorli** [comment]
Found and fixed the issue in the "Add GetLoaderAllocator to ICLRPrivBinder" commit. 
@davidwrighton can you please take a look and approve the PR if you are fine with it now?

**Comment by masonwheeler** [comment]
Woohoo! :tada:

**Comment by TorVestergaard** [comment]
Awesome job! So, what does this PR going through and being merged mean for the potential future release of AssemblyLoadContext unloading, exactly?

**Comment by janvorli** [comment]
@TorVestergaard this is the first step towards that goal. With the merged-in state, I am able to run coreclr tests using a wrapper tool that loads a test and its dependencies into an AssemblyLoadContext and unloads it after the test finishes. 
There are still some tests that don't pass in such an environment due to four limitations that the unloadability now has. This is their list in the order of priority:
* There can be no thread local member or static variables in the assemblies loaded into the unloadable AssemblyLoadContext
* Delegate marshaling for types within collectible assemblies is not supported.
* COM Interop is not supported for collectible types
* FixedAddressValueTypeAttribute is not supported for fields in collectible types
If an attempt to load assemblies with these limitations into an unloadable AssemblyLoadContext is made, an exception is thrown.

I am currently working on investigating these limitations with the ultimate goal to get rid of them. 

We also need more testing and also testing more complex scenarios like multiple AssemblyLoadContexts coexisting and interacting, etc.

**Comment by jkotas** [comment]
We also need:
- Expose the collectible option in the public API surface: https://github.com/dotnet/corefx/issues/14724
- Add related APIs that make it possible for libraries to do the right thing for collectible assemblies like: https://github.com/dotnet/corefx/issues/25671 .
